examples/conv_filter_visualization.py

```diff
@@ -41,99 +53,189 @@ def deprocess_image(x):
+    '''
     if K.image_data_format() == 'channels_first':
-        input_img_data = np.random.random((1, 3, img_width, img_height))
-    else:
-        input_img_data = np.random.random((1, img_width, img_height, 3))
-    input_img_data = (input_img_data - 0.5) * 20 + 128
-
-    # we run gradient ascent for 20 steps
-    for i in range(20):
-        loss_value, grads_value = iterate([input_img_data])
-        input_img_data += grads_value * step
-
-        print('Current loss value:', loss_value)
-        if loss_value <= 0.:
-            # some filters get stuck to 0, we can skip them
-            break
-
-    # decode the resulting input image
-    if loss_value > 0:
-        img = deprocess_image(input_img_data[0])
-        kept_filters.append((img, loss_value))
-    end_time = time.time()
-    print('Filter %d processed in %ds' % (filter_index, end_time - start_time))
-
-# we will stich the best 64 filters on a 8 x 8 grid.
-n = 8
-
-# the filters that have the highest loss are assumed to be better-looking.
-# we will only keep the top 64 filters.
-kept_filters.sort(key=lambda x: x[1], reverse=True)
-kept_filters = kept_filters[:n * n]
-
-# build a black picture with enough space for
-# our 8 x 8 filters of size 128 x 128, with a 5px margin in between
-margin = 5
-width = n * img_width + (n - 1) * margin
-height = n * img_height + (n - 1) * margin
-stitched_filters = np.zeros((width, height, 3))
-
-# fill the picture with our saved filters
-for i in range(n):
-    for j in range(n):
-        img, loss = kept_filters[i * n + j]
-        width_margin = (img_width + margin) * i
-        height_margin = (img_height + margin) * j
-        stitched_filters[
-            width_margin: width_margin + img_width,
-            height_margin: height_margin + img_height, :] = img
-
-# save the result to disk
-save_img('stitched_filters_%dx%d.png' % (n, n), stitched_filters)
+        x = x.transpose((2, 0, 1))
+    return (x / 255 - 0.5) * 4 * former.std() + former.mean()
+
+
+def visualize_layer(model,
+                    layer_name,
+                    step=1.,
+                    epochs=10,
+                    upscaling_steps=9,
+                    upscaling_factor=1.2,
+                    output_dim=(412, 412)):
+    '''Visualizes the most relevant filters of one conv-layer in a certain model.
+
+    # Arguments
+        model: The model containing layer_name.
+        layer_name: The name of the layer to be visualized.
+                    Has to be a part of model.
+        step: step size for gradient ascent.
+        epochs: Number of iterations for gradient ascent.
+        upscaling_steps: Number of upscaling steps.
+                         Starting image is in this case (80, 80).
+        upscaling_factor: Factor to which to slowly upgrade
+                          the image towards output_dim.
+        output_dim: [img_width, img_height] The output image dimensions.
+    '''
+
+    def _generate_filter_image(input_img,
```

Comment:
Of course, they can live outside, but they're not usable from outside. Specifically, `visualize_layer` enforces the right usage and computation of the inputs of those two functions.

Reply:
If nested functions don't conform to your codestyling ideas, I could possibly transform `viszalize_layer` into a class. However, in my point of view, I'd come to the conclusion that a class would be too overdressed especially since this should be an easily readable example!